bash_completion

```diff
@@ -1148,6 +1148,17 @@ _comp_compgen_filedir()
         compopt -o filenames 2>/dev/null
     fi
 
+    # Remove . and .. (as well as */. and */..) from suggestions,
+    # unless .. or */.. was typed explicitly by the user
+    # (for users who use tab-completion to append a slash after '..')
+    if [[ "${cur}" != @(..|*/..) ]]; then
+        local i
+        for i in "${!toks[@]}" ; do
+            [[ "${toks[$i]}" == @(.|..|*/.|*/..) ]] && \
+                unset -v "toks[$i]"
+        done
```

Comment:
```suggestion
        if ((${#toks[@]})); then
            _comp_compgen -Rv toks -- -X '@(.|..|*/.|*/..)' -W '"${toks[@]}"'
        fi
```

In addition, I think we can return from the function when `((${#toks[@]} == 0))` before the above `compopt -o filenames 2>/dev/null` (line 1148). Then, the check for `((${#toks[@]}))` can be skipped.

Reply:
This filtering should probably be done before setting `compopt -o filenames`.

Reply:
1843ad5d2